frontend/lib/resolution/call-init-deinit.cpp

```diff
@@ -94,35 +100,373 @@ struct CallInitDeinit : VarScopeVisitor {
+    return false;
+  } else if (t->isEnumType()) {
+    // OK, can default-initialize enums to first element
+    return false;
+  } else if (t->isFunctionType()) {
+    return false;
+  } else if (auto ct = t->toClassType()) {
+    auto decorator = ct->decorator();
+    // no action needed for 'borrowed' or 'unmanaged'
+    // (these should just default initialized to 'nil',
+    //  so nothing else needs to be resolved)
+    if (decorator.isBorrowed() || decorator.isUnmanaged()) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+
+void CallInitDeinit::resolveDefaultInit(const VarLikeDecl* ast, RV& rv) {
+
+  ResolvedExpression& varRes = rv.byAst(ast);
+  QualifiedType varType = varRes.type();
+
+  if (varType.isErroneousType()) {
+    // ignore it -- should have already issued an error
+    return;
+  }
+  if (varType.isUnknownKindOrType()) {
+    context->error(ast, "cannot default initialize variable using generic or unknown type");
+    return;
+  }
+  // check genericity
+  auto g = getTypeGenericity(context, varType);
+  if (g != Type::CONCRETE) {
+    context->error(ast, "cannot default initialize variable using generic or unknown type");
+    return;
+  }
+
+  // these will be set below
+  const ClassType* classType = nullptr;
+  const CompositeType* compositeType = nullptr;
+
+  const Type* vt = varType.type();
+
+  if (auto ct = vt->toClassType()) {
+    auto decorator = ct->decorator();
+    // check that the class is a nilable class type
+    if (!decorator.isNilable()) {
+      // TODO: improve this error
+      context->error(ast, "cannot default initialize variable using non-nilable class type");
+      return;
+    }
+  }
+
+  if (!typeNeedsInitDeinitCall(vt)) {
+    // nothing to do here
+  } else if (vt->isTupleType()) {
+    // TODO: probably need to do something here, at least in some cases
+  } else if (auto ct = vt->toClassType()) {
+    auto decorator = ct->decorator();
+    // check that the class is a nilable class type
+    if (!decorator.isNilable()) {
+      // TODO: improve this error
+      context->error(ast, "cannot default initialize variable using non-nilable class type");
+      return;
+    }
+    // no action needed for 'borrowed' or 'unmanaged'
+    // (these should just default initialized to 'nil',
+    //  so nothing else needs to be resolved)
+    if (decorator.isBorrowed() || decorator.isUnmanaged()) {
+      return;
+    }
+
+    // otherwise, need to resolve an 'init' e.g. shared.init
+    classType = ct;
+    compositeType = ct->manager()->toCompositeType();
+  } else if (auto ct = vt->toCompositeType()) {
+    compositeType = ct;
+  }
+
+  if (compositeType != nullptr) {
+    // try to resolve 'init'
+    // TODO: handle instantiations passing field types
+    std::vector<CallInfoActual> actuals;
+    actuals.push_back(CallInfoActual(varType, USTR("this")));
+    if (classType != nullptr && classType->manager() != nullptr) {
+      // always pass chpl_t=borrowed class type
```

Comment:
Not sure I'm following the logic here; could the comment elaborate a little bit?

Reply:
I've improved the comment.